Added back all PMA code

Dr15Jones opened the pull request.

Dr15Jones committed:
Added back all PMA code

Dr15Jones commented:
@erica-snider @kutschke @marcpaterno 
Here is the artificial pull request with all the PMA code

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaElement3D.h:
```diff
@@ -0,0 +1,118 @@
+	double fSumHitsQ[3];
+	double fHitsRadius;
+
+	static float fOptFactors[3]; // impact factors of data from various 2D views
```
Looks like this could be `constexpr`

Dr15Jones commented:
@robertsulej Here is the artificial pull request

Dr15Jones commented:
In order to help me better understand the relationships between the various classes used in PMA, I created a UMLish diagram showing their relationships based on member data

[LArSoft PMA Class Relationships.pdf](https://github.com/larsoft-test/larreco/files/250905/LArSoft.PMA.Class.Relationships.pdf)

kutschke commented:
Hi Chris,

  This is the most recent that I have received.

---

From: Chris Jones notifications@github.com
Sent: Thursday, May 5, 2016 1:02:58 PM
To: larsoft-test/larreco
Cc: Robert K Kutschke; Mention
Subject: Re: [larsoft-test/larreco] Added back all PMA code (#2)

In order to help me better understand the relationships between the various classes used in PMA, I created a UMLish diagram showing their relationships based on member data

LArSoft PMA Class Relationships.pdfhttps://github.com/larsoft-test/larreco/files/250905/LArSoft.PMA.Class.Relationships.pdf

## 

You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHubhttps://github.com/larsoft-test/larreco/pull/2#issuecomment-217227974

Dr15Jones commented:
@kutschke that's perfect. Thanks for checking.

Dr15Jones commented:
I had missed some member data to the central EDProducer. This is the revised diagram.

[LArSoft PMA Class Relationships.pdf](https://github.com/larsoft-test/larreco/files/250994/LArSoft.PMA.Class.Relationships.pdf)

Dr15Jones committed:
Split PMAlgTrackMaker_module.cc into two files

Github has a limit of 1500 lines for one file when using the diff tool
and PMAlgTrackMaker_module.cc exceeded 2500. Therefore to aid in the
review the file was split into to parts.

Dr15Jones commented:
@amundson this should cause you to get emails for all comments to this artificial pull request.

erica-snider commented on larreco/RecoAlg/PMAlg/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+			${MF_UTILITIES}
+	        	${FHICLCPP}
+			${CETLIB}
+            		${CLHEP}
```
hello

kutschke commented on larreco/RecoAlg/ProjectionMatchingAlg.cxx:
```diff
@@ -0,0 +1,1151 @@
+
+	std::vector< unsigned int > trkTPCs = trk.TPCs();
+	std::vector< unsigned int > trkCryos = trk.Cryos();
+	std::map< std::pair< unsigned int, unsigned int >, std::pair< TVector2, TVector2 > > ranges;
```
This is my first attempt at commenting - please let me know if you see it.

1)
LArSoft should make a global decision about standard 2-vector, 3-vector, 4-vector classes (and the classes that perform transformations on them).  When making this decision, be aware that a an object of type TVector2 is also a TObject and has a size of 32 bytes - 16 bytes of (x,y) and 16 bytes of TObject-ness.

2)
In this line there are two uses of std::pair. I think that this code would be easier to understand if these are replaced by two well named structs each with two well named elements.  

3)
It would be a good idea if the geometry subsystem provided a 0-based dense integer identifier for all of the tpc+cryo in the system.  Then the std::map could be replaced by a std::vector.

robertsulej replied:
Hello,
Arrived. I can see this and following comments, thanks.
R.

kutschke replied:
Hi Robert,

  I hope that your trip went well.  See you and Dorota tomorrow.

  My plan is to write down every comment that comes to mind even fairly minor ones.  This way they don’t get lost.   We can worry about priorities later.  Of course some might be rejected.

```
  Rob
```

> On May 9, 2016, at 6:56 PM, Robert Sulej notifications@github.com wrote:
> 
> In larreco/RecoAlg/ProjectionMatchingAlg.cxx:
> 
> > -   pma::Element3D::SetOptFactor(geo::kV, p.get< double >("HitWeightV"));
> > -   pma::Element3D::SetOptFactor(geo::kU, p.get< double >("HitWeightU"));
> >   +}
> >   +// ------------------------------------------------------
> >   +
> >   +double pma::ProjectionMatchingAlg::validate(const pma::Track3D& trk,
> > -   const std::vector< art::Ptrrecob::Hit >& hits,
> > -   unsigned int testView) const
> >   +{
> > -   double max_d = fTrkValidationDist2D;
> > -   double d2, max_d2 = max_d \* max_d;
> > -   unsigned int nAll = 0, nPassed = 0;
> >   +
> > -   std::vector< unsigned int > trkTPCs = trk.TPCs();
> > -   std::vector< unsigned int > trkCryos = trk.Cryos();
> > -   std::map< std::pair< unsigned int, unsigned int >, std::pair< TVector2, TVector2 > > ranges;
> 
> Hello,
> Arrived. I can see this and following comments, thanks.
> R.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+void PMAlgTrackMaker::reconfigure(fhicl::ParameterSet const& pset)
+{
+	fHitModuleLabel = pset.get< std::string >("HitModuleLabel");
+	fCluModuleLabel = pset.get< std::string >("ClusterModuleLabel");
```
Consider making these art::InputTag rather than std::string; this allows users to specify by module label and instance name in one place.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module_continued.h:
```diff
@@ -0,0 +1,1039 @@
+
+	fDetProp = lar::providerFrom<detinfo::DetectorPropertiesService>();
+
+	reset(evt); // set default values, clear containers at the beginning of each event
```
Do you also clear at end of event?  If not you should.  It's probably best to do both beginning and end of event.  At end to reduce peak memory usage and at beginning in case art continued after a throw from within your module ( which may never happend today but may in the future ).

marcpaterno replied:
I think it would be good to consider removing this from the module state, and to have the objects in question be function-scope -- so that they are destroyed upon exit from the function. This guarantees the data do not linger in memory longer than needed, and provides exception safety for free.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+	fTree_trk->Branch("fLength", &fLength, "fLength/D");
+	fTree_trk->Branch("fHitsMse", &fHitsMse, "fHitsMse/D");
+	fTree_trk->Branch("fSegAngMean", &fSegAngMean, "fSegAngMean/D");
+	fTree_trk->Branch("fPidTag", &fPidTag, "fPidTag/I");
```
Does LArSoft have a policy about producer modules making TTrees?  If this TTree is just a duplicate of information available in the data products then I suggest moving the TTree creation and filling to a standalone analyzer module.  If it contains diagnostic information not present in the data products there should be build-time switches to build a version of this code that is free of TTree - in production mode you do not want pay any memory of CPU price related to diagnostics.  If the TTree contains analysis level information not available in the data products, check with LArSoft management if this is permitted.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+typedef std::map< size_t, std::vector<double> > dedx_map;
+typedef std::map< unsigned int, std::vector< art::Ptr<recob::Hit> > > view_hitmap;
+typedef std::map< unsigned int, view_hitmap > tpc_view_hitmap;
+typedef std::map< unsigned int, tpc_view_hitmap > cryo_tpc_view_hitmap;
```
Suggest that the experts comment on the design of this data structure.  It's essentially a three index container of std::vectorart::Ptr<recob::Hit; the three indices are IDs for a cryostat, a TPC within a cryostat and a view within a  TPC.   All three indices are always small (or smallish) integers. 
1. Is std::map appropriate for the outer three containers?  
2. Is the right answer to invent a dense integer ID for the (cryo,tpc,view) tuple and then use a std::vector as the outer container?  ( an earlier version of this comment had std::array but the dimensions are only known at run time)
3.  If so, the dense ID could be defined here but it would be better if it were defined in the geometry subsystem; this involves cooperation across multiple experiments - is it practical?
4.  In this data structure, do all Ptr's always point into a single data product - or is it possible that they may point into more than one data product?  If the former then switch to bare  pointers and keep track of the product ID separately. This will have both speed and memory footprint improvements.
5. If all Ptrs point to a single data product, does it make sense to design that data product to support the 3-index structure used here?  Could other algorithms use such a reorganization?  Work with other algorithm developers to understand how to structure the raw hit containers to support the algorithms.
6. The code would be easier for next-generation maintainer/developer to follow if you have types for the cryo ID, tic ID and view ID.  Just a typedef helps; doing the work get develop an opaque integer will turn crossed-indexing into a compile time error.

kutschke replied:
A followup to my note above.  If you need to support the possibility that the hit Ptr's point to objects in more than one data product, I see two classes of solutions
1. If the data products are all std::vectors, then the transient representation can be a a bare pointer.  We would need to provide the appropriate helper to reconstitute Ptrs.  This works for any data product in which the elements are contiguous.
2. Another solution is to define a "transient Ptr" class that has the persistent info plus a bare pointer; it would not do any validity check so its as fast to use as a bare pointer. Again we would have to add helpers to reconstitute Ptrs.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+#include <memory>
+
+namespace trkf {
+
```
Should the following typedefs be at namespace scope or at class scope?

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaElement3D.cxx:
```diff
@@ -0,0 +1,342 @@
+float pma::Element3D::fOptFactors[3] =     { 0.2F, 0.8F, 1.0F };
+
+pma::Element3D::Element3D() :
+	fTPC(-1), fCryo(-1),
```
I'd suggest that instead of using `-1` explicitly, create symbolic constants,  e.g. `kInvalidTPC` and `kInvalidCryo`, which you set to `-1`. Then user code could use those values and it would help readability of the code. In addition, it would be possible to change the representation of the value (e.g. become `unsigned`) and using symbolic names would keep code working.

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaElement3D.cxx:
```diff
@@ -0,0 +1,342 @@
+	fHitsRadius(0)
+{
+	fNThisHitsEnabledAll = 0;
+	for (unsigned int i = 0; i < 3; i++)
```
Instead of using `3` explicitly, it would help to create a symbolic constant set to 3 and then use it for the array sizes as well as to iterate over the elements.
Another option to consider is to switch to `std::array<>` since it would initialize the values to 0 and allows standard iteration.

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaElement3D.cxx:
```diff
@@ -0,0 +1,342 @@
+{
+	size_t n = 0;
+	for (size_t i = 0; i < fAssignedHits.size(); i++)
+		if (fAssignedHits[i]->IsEnabled() &&
```
Given the index value `i` is not needed for this loop, I'd suggest switching to the `for( auto const& hit: fAssignedHits)` form.
In addition, I suggest adding `{...}` to clarify scope. I know this code presently works, but if it is ever modified it is very easy to think the modification is within the `for` or `if` scope but be wrong.

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaHit3D.h:
```diff
@@ -0,0 +1,104 @@
+
+	unsigned int Cryo(void) const { return fCryo; }
+	unsigned int TPC(void) const { return fTPC; }
+	unsigned int View2D(void) const { return fPlane; }
```
From the code, is it correct to say that the only valid values are `geo::Z`, `geo::U` and `geo::V`? If so, how about changing this to return an enum?

marcpaterno commented on larreco/TrackFinder/PMAlgTrackMaker_module_continued.h:
```diff
@@ -0,0 +1,1039 @@
+
+	pma::TrkCandidateColl result;
+
+	std::unique_ptr< std::vector< recob::Track > > tracks(new std::vector< recob::Track >);
```
Consider replacing
  std::unique_ptrstd::vector<recob::Track> tracks(new std::vectorrecob::Track);
with
  auto tracks = std::make_uniquestd::vector<recob::Track>();
and similarly for all other unique_ptr instances below.

Dr15Jones commented on larreco/RecoAlg/PMAlg/PmaElement3D.cxx:
```diff
@@ -0,0 +1,342 @@
+	fNThisHits[1] = hitsInd2.size();
+	fNThisHits[2] = hitsColl.size();
+
+	pma::SortedObjectBase const * chain = dynamic_cast< pma::SortedObjectBase* >(this);
```
I think reviewing the use of dynamic_cast here might be a worthwhile activity.

Dr15Jones commented on larreco/RecoAlg/ProjectionMatchingAlg.cxx:
```diff
@@ -0,0 +1,1151 @@
+	const std::vector< art::Ptr<recob::Hit> >& hits_1,
+	const std::vector< art::Ptr<recob::Hit> >& hits_2) const
+{
+	pma::Track3D* trk = new pma::Track3D(); // track candidate
```
Would be good to consider using `std::unique_ptr<>` to manage the memory.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+	fMakePFPs = pset.get< bool >("MakePFPs");
+
+	fMinSeedSize1stPass = pset.get< size_t >("MinSeedSize1stPass");
+	fMinSeedSize2ndPass = pset.get< size_t >("MinSeedSize2ndPass");
```
It would be a good idea to validate inputs.  There are new features coming in art to assist with this.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+  // ******************** fcl parameters **********************
+  std::string fHitModuleLabel; // label for hits collection (used for trk validation)
+  std::string fCluModuleLabel; // label for input cluster collection
+  int fCluMatchingAlg;         // which algorithm for cluster association
```
Consider defining an enum to enumerate all valid values.  Make the data member of enum type.

kutschke commented on larreco/TrackFinder/PMAlgTrackMaker_module.cc:
```diff
@@ -0,0 +1,1368 @@
+{
+	fHitModuleLabel = pset.get< std::string >("HitModuleLabel");
+	fCluModuleLabel = pset.get< std::string >("ClusterModuleLabel");
+	fCluMatchingAlg = pset.get< int >("CluMatchingAlg");
```
See the comment above about making fCluMatchingAlg and enum type;  also consider making the parameter value a string that matches the name of the enum.  We can point you to some tools to help turn  strings into enum types.  The tools also validate the input.

kutschke commented on larreco/RecoAlg/PMAlg/PmaTrack3D.h:
```diff
@@ -0,0 +1,318 @@
+	pma::Element3D* GetNearestElement(const TVector2& p2d, unsigned int view, int tpc = -1,
+		bool skipFrontVtx = false, bool skipBackVtx = false) const;
+	pma::Element3D* GetNearestElement(const TVector3& p3d) const;
+	std::vector< pma::Node3D* > fNodes;
```
We understand that nodes can be shared among tracks.  Suggest that they be held by shared ptr, not bare pointer.

kutschke commented on larreco/RecoAlg/PMAlg/PmaTrack3D.h:
```diff
@@ -0,0 +1,318 @@
+		bool skipFrontVtx = false, bool skipBackVtx = false) const;
+	pma::Element3D* GetNearestElement(const TVector3& p3d) const;
+	std::vector< pma::Node3D* > fNodes;
+	std::vector< pma::Segment3D* > fSegments;
```
We initially thought that tracks segments could be owned by tracks - so there was a comment here about changing the std::vector to hold segments by value, not pointer.  We now realize that this is not quite right - a segment might get moved from one track to another.  So suggest that the segments be held by shared ptr, not bare pointer.

kutschke commented on larreco/RecoAlg/PMAlg/PmaSegment3D.cxx:
```diff
@@ -0,0 +1,276 @@
+}
+
+double pma::Segment3D::GetDist2(const TVector2& psrc, const TVector2& p0, const TVector2& p1)
+{
```
Reminder about recommendation to find a replacement for TVector2.

The algorithm expressed in this function can be simplified by doing some more algebra by hand.

Dr15Jones replied:
Given the time critical nature of this calculation, doing it directly from the vector components would probably be beneficial. The formula is given in 
http://mathworld.wolfram.com/Point-LineDistance2-Dimensional.html
